tests/test_repro.py

```diff
@@ -1027,3 +1028,30 @@ def test(self):
         shutil.rmtree(self.dvc.scm.dir)
         ret = main(['repro', self.file1_stage])
         self.assertEqual(ret, 0)
+
+
+class TestReproAllPipelines(TestDvc):
+    def test(self):
+        self.dvc.run(fname='start.dvc',
+                     outs=['start.txt'],
+                     cmd='echo start > start.txt')
+
+        self.dvc.run(fname='middle.dvc',
+                     deps=['start.txt'],
+                     outs=['middle.txt'],
+                     cmd='echo middle > middle.txt')
+
+        self.dvc.run(fname='final.dvc',
+                     deps=['middle.txt'],
+                     outs=['final.txt'],
+                     cmd='echo final > final.txt')
+
+        self.dvc.run(fname='disconnected.dvc',
+                     outs=['disconnected.txt'],
+                     cmd='echo other > disconnected.txt')
+
+        mock_reproduce = patch.object(Stage, 'reproduce').start()
```

Comment:
One more thing. You are leaking mock here, this is why all the tests are randomly failing. Please use `with patch.object(Stage, 'reproduce') as mock_reproduce:` here.